### ISSUE TITLE
Is there any reason for `TimeDistributedDense`?

### ISSUE THREAD

**Issue by wxs** [fchollet/keras#278]
Similar to how I removed time_distributed_softmax, and just made it always apply softmax to the last axis (i.e. the nb_dimensions) axis, is there any reason that I don't make `Dense` have the same behavour, so you can pass it either a (nb_samples, nb_dims) or a (nb_samples, nb_timesteps, nb_dims) matrix? Seems like it reduces complexity with no downside.

**Comment by wxs** [comment]
This would also be more efficient because the current `TimeDistributedDense` implementation uses a `scan` for some reason, though it should work without one.

**Comment by phreeza** [comment]
Possibly for the case when the entire data doesn't fit into GPU memory?

**Comment by fchollet** [comment]
> This would also be more efficient because the current TimeDistributedDense implementation uses a scan for some reason, though it should work without one.

That has actually been performance-checked at the time, and both direct matrix multiplication and scan had the exact same performance. 

I think it might be clearer to keep both Dense and TimeDistributedDense, since "Dense" or "fully-connected" implies that all input values feed into a single weight matrix. With TimeDistributedDense however, there are no temporal connections. 

I'll leave this to what the majority of users think. If there is no risk of confusion, then we can definitely switch to a single Dense layer.

**Comment by wxs** [comment]
Incidentally as of #283 TimeDistributedDense is now implemented identically to Dense (except for a dimshuffle which I think can be generalized)

**Comment by the-moliver** [comment]
I actually side with keeping it separate for the same reason as fchollet: it is conceptually distinct from a fully dense layer.

**Comment by wxs** [comment]
Not sure I agree: *none* of the layers have temporal connections except for the recurrent ones (GRU, LSTM, SimpleRNN, etc.). In all cases the expected behaviour when given temporal data is to operate identically at each timestep (this is how all the activations work, for instance).

The "natural" behaviour of a layer is to not be able to see back in time, and recurrent layers are the exception.

Plus: though you write that Dense "implies that all input values feed into a single weight matrix", it doesn't actually in its current incarnation operate with connections across time. You would need to first pass the time data through a Reshape. Or modify Dense as follows:

     def get_output(self, train):
         X = self.get_input(train)
         X = T.flatten(X, ndim=2)
         output = self.activation(T.dot(X, self.W) + self.b)
         return output    

I think this would be a more confusing way to interpret Dense, though, and wouldn't work with variable length inputs, which would be inconsistent with the rest of the package.

**Comment by the-moliver** [comment]
Hmm, fair points. However, I'm actually planning to introduce connections across time (basically a 1D convolution) into TimeDistributedDense (to enable time-delay neural networks) which was my main reason for wanting to keep it as a separate layer type. It would work such that when set to have 0 delays it would function as it does now.

**Comment by wxs** [comment]
Ah: yes if `TimeDistributedDense` did a time-convolution that would be interesting.

I do think that previously the naming convention "TimeDistributedX" layers was that they did *not* look at the time dimension. But given that now TimeDistributedDense is the only such surviving layer I guess if its behaviour changes there is no one else following the convention to be updated.

Though doesn't Convolution1D already let you do time convolutions? I haven't used it but it is documented as doing that..

**Comment by the-moliver** [comment]
You may very well be right and in which case I'm fine with getting rid of TimeDistributedDense. I haven't used Convolution1D yet either but will test it to see if it already does what I want.

**Comment by falaktheoptimist** [comment]
So, from this discussion I glean that:
`model.add(LSTM(512, input_shape=(maxlen, len(chars)),return_sequences=True)) 
model.add(LSTM(512, return_sequences=True)) #- original
model.add(Dropout(0.2))
model.add(Dense(len(chars)))
model.add(Activation('softmax'))
`

and 

`model.add(LSTM(512, input_shape=(maxlen, len(chars)),return_sequences=True)) 
model.add(LSTM(512, return_sequences=True)) #- original
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(len(chars))))
model.add(Activation('softmax'))
`

represent the exact same model. I also checked the summary and the number of parameters are also exactly the same which leads me to conclude that I can probably do away with TimeDistribute block altogether without affecting the model in any way. Please correct me here if this is incorrect

**Comment by DomHudson** [comment]
@falaktheoptimist 
Did you ever work this out? I'm wondering the same thing - the output dimensions, loss and accuracy are identical for me if I replace my final `TimeDistributed(Dense())` layer with just a `Dense` layer.
